### PR TITLE
Adding some more functionality to `bcf::Record`

### PR DESCRIPTION
**Pull request by holtgrewe**

### CODE REVIEW

File: src/bcf/record.rs [hunk 1]
```diff
@@ -140,6 +170,41 @@ impl Record {
         (0..n).map(|i| unsafe { ffi::CStr::from_ptr(alleles[i]).to_bytes() }).collect()
     }
 
+    /// Update alleles as vector of alleles.
+    pub fn update_alleles(&mut self, alleles: &Vec<&[u8]>) -> Result<(), AlleleWriteError> {
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply src/bcf/record.rs]: The signature should be
```
/// Set alleles. Alleles can be given as anything that implements `AsRef<[u8]>`, i.e., 
/// can be converted into a slice of `u8`. For example, a vector of slices, vector of vectors, 
/// slices of slices, and so on.
pub fn update_alleles<A: AsRef<[u8]>>(&mut self, alleles: &[A]) -> Result<(), AlleleWriteError> {
```

File: src/bcf/record.rs [hunk 2]
```diff
@@ -140,6 +170,41 @@ impl Record {
...
+                ffi::CString::new(alleles).unwrap().as_ptr() as *mut i8,
+            ) == 0 {
+                Ok(())
+            } else {
+                Err(AlleleWriteError::Some)
+            }
+        }
+    }
+    // int bcf_update_alleles(const bcf_hdr_t *hdr, bcf1_t *line, const char **alleles, int nals);
+    // int bcf_update_alleles_str(const bcf_hdr_t *hdr, bcf1_t *line, const char *alleles_string);
```
Comment (johanneskoester) [inline src/bcf/record.rs]: Please remove above two lines.

File: src/bcf/record.rs [hunk 3]
```diff
@@ -140,6 +170,41 @@ impl Record {
...
+            ) == 0 {
+                Ok(())
+            } else {
+                Err(AlleleWriteError::Some)
+            }
+        }
+    }
+
+    /// Update alleles as comma-separated ASCII string of alleles.
+    pub fn update_alleles_str(&mut self, alleles: &[u8]) -> Result<(), AlleleWriteError> {
```
[the remark this answers is not in the document]
Comment (johanneskoester) [reply src/bcf/record.rs]: Do we really want to expose this? I don't think usage of this variant should be encouraged.

File: src/bcf/record.rs [hunk 4]
```diff
@@ -191,6 +256,12 @@ impl Record {
         self.push_format(tag, data, htslib::BCF_HT_REAL)
     }
 
+    /// Update genotypes for all samples, corresponds to `bcf_update_genotypes` in HTSlib.
+    pub fn push_format_genotypes(&mut self, gts: &[i32]) -> Result<(), TagWriteError> {
```
Comment (johanneskoester) [inline src/bcf/record.rs]: Maybe this should rather be `set_genotypes(&mut self, gts: &[Genotypes])` and take a [Genotypes](https://docs.rs/rust-htslib/0.17.0/rust_htslib/bcf/record/struct.Genotypes.html) object? I know, this requires some additional implementation work.

File: src/bcf/mod.rs [hunk 5]
```diff
@@ -21,6 +21,11 @@ pub use bcf::record::Record;
 pub use bcf::buffer::RecordBuffer;
 
 
+/// Redefinition of corresponding `#define` in `vcf.h.`.
+#[allow(non_upper_case_globals)]
+pub const gt_missing: i32 = 0;
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/mod.rs]: OK

File: src/bcf/record.rs [hunk 6]
```diff
@@ -131,6 +131,36 @@ impl Record {
         self.inner_mut().pos = pos;
     }
 
+    /// Update the ID string to the given value.
+    pub fn update_id(&mut self, id: &[u8]) -> Result<(), IdWriteError> {
+        unsafe {
+            if htslib::bcf_update_id(
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/record.rs]: OK

File: src/bcf/record.rs [hunk 7]
```diff
@@ -131,6 +131,36 @@ impl Record {
...
+            ) == 0 {
+                Ok(())
+            } else {
+                Err(IdWriteError::Some)
+            }
+        }
+    }
+
+    /// Add the ID string (the ID field is semicolon-separated), checking for duplicates.
+    pub fn add_id(&mut self, id: &[u8]) -> Result<(), IdWriteError> {
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/record.rs]: OKD

File: src/bcf/record.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/record.rs]: OK

File: src/bcf/record.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (holtgrewe) [reply src/bcf/record.rs]: OK

File: src/bcf/record.rs (hunk 2 above)
Comment (holtgrewe) [reply src/bcf/record.rs]: OK

File: src/bcf/record.rs (hunk 4 above)
Comment (holtgrewe) [reply src/bcf/record.rs]: I'm removing this for now, adding as Issue in #67.

File: src/bcf/record.rs [hunk 8]
```diff
@@ -140,6 +166,26 @@ impl Record {
         (0..n).map(|i| unsafe { ffi::CStr::from_ptr(alleles[i]).to_bytes() }).collect()
     }
 
+    /// Set alleles. Alleles can be given as anything that implements `AsRef<[u8]>`, i.e.,
+    /// can be converted into a slice of `u8`. For example, a vector of slices, vector of vectors,
+    /// slices of slices, and so on.
+    pub fn update_alleles<'a, A: AsRef<[u8]>>(&mut self, alleles: &'a [A]) -> Result<(), AlleleWriteError>
```
Comment (johanneskoester) [inline src/bcf/record.rs]: Can you also rename this to `set_alleles`?

Comment (holtgrewe) [reply src/bcf/record.rs]: Done.

**Pull request merged**